Sources/Instrumentation/Tracing/Span.swift

```diff
@@ -0,0 +1,113 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Tracing open source project
+//
+// Copyright (c) 2020 Moritz Lang and the Swift Tracing project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Baggage
+import Dispatch
+
+public protocol Span {
+    var operationName: String { get }
+
+    var startTimestamp: DispatchTime { get }
+    var endTimestamp: DispatchTime? { get }
+
+    var baggage: BaggageContext { get }
+
+    var events: [SpanEvent] { get }
+    mutating func addEvent(_ event: SpanEvent)
+
+    var attributes: [String: SpanAttribute] { get }
```

Comment:
> Attributes SHOULD preserve the order in which they're set. Setting an attribute with the same key as an existing attribute SHOULD overwrite the existing attribute's value.

implies we might want to not expose a Dictionary here, we can address this in a separate ticket; There's no good sorted dict today sadly; We could consider a struct to hide the dict.

Reply:
Hiding behind a struct sounds useful, then we could solve this via an implementation detail. Maybe an array of tuples like `HTTPHeaders` where instead of simply appending new tuple values updating the existing one with the same key. Might be a performance hit though.

Reply:
Sometimes scanning through an array can be better than a dictionary :) It depends on amount of the data and how it's used; It can be good for us to hide what we're using to implement this -- then we can change around and noone will notice.

Reply:
Yes, let's resolve this for now by hiding the dictionary behind a struct. 👍